design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,264 @@
+
+- The JSON structure is validated and enforced. If the input is invalid, the user will get a proper error message and the variable update will be denied.
+- The JSON structure will not be encrypted completely, but only the sensitive information. This will allow future seach capabilities based on the non sensitive information.
+
+### User Experience
+
+Let's have a look at the following example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+  - !variable
+    id: mysql-db-creds
+    schema: !variable /conjur/schemas/mysql-schema
+
+  - !host my-app
+  
+  - !permit
+    role: !host my-app
+    privileges: [ read,execute ]
+    resource: !variable mysql-db-creds
+```
+
+And
+
+```yaml
+- !policy
+  id: conjur/schemas
+  body:
+  - &schemas
+    - !variable mysql-schema
+  
+  - !permit
+    role: !group /conjur/all
+    privileges: [ read,execute ]
+    resource: !variable *schemas
+```
+
+In the example above, the `/conjur/schemas/mysql-schema` variable will be updated with the following value:
+
+```json
+{
+  "$id": "https://cyberark.com/mysql.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "MySQL Creds",
+  "type": "object",
+  "properties": {
+    "username": {
+      "type": "string",
+      "description": "The MySQL username"
+    },
+    "password": {
+      "type": "string",
+      "description": "The MySQL password",
+      "minLength": 12,
+      "maxLength": 40,
+      "pattern": "[A-Za-z0-9]"
+    },
+    "address": {
+      "type": "string",
+      "description": "The MySQL address"
+    },
+    "port": {
+      "description": "The MySQL port number",
+      "type": "integer",
+      "minimum": 1,
+      "maximum": 65535
+    }
+  },
+  "required": ["username", "password"],
+  "additionalProperties": false,
+  "secrets": [
+      "password",
+    ]
+}
+```
+
+And the `mysql-db-creds` variable could be updated with the following value:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password-12345",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+### Functionality
+
+#### Input Validation
+
+Conjur will validate the variable content, using the JSON schema that is specified in the `schema` attribure of that variable.
+
+For the example above, Conjur will use the JSON schema specified in `conjur/schemas/mysql-schema` to verify the content of `mysql-db-creds`. This verification includes:
+
+- `username` and `password` are mandatory properties.
+- `address` and `port` are allowed properties.
+- `port` must contain a valid port number.
+- `password` must align to a sufficient complexity.
+
+If the varialbe content is invalid, the user will get an HTTP code of 422 (Unprocessable Entity) and the body of the response will also contain an elaborative error message that explains what part of the input was found to be invalid.
+
+#### Secrets Encryption
```

Comment:
If I understand this section correctly, my perspective here is that per-field/attribute encryption of credentials here is unnecessary if the whole variable value is already encrypted in the database at-rest. I don't think we gain much by doing a per-field encryption and would increase code complexity by quite a bit for it.

Reply:
You raise a good point, that I don't think I explained well enough. The proposal suggests that the variable will no longer be encrypted as a whole, but only the secret field would, while the JSON structure and the non sensitive data will remain clear. This will help support search capabilities, for example. Look for a password, based on the username. But also, this is not a minimal requirement, this proposal could also be implemented without this feature, and we could leave that for later and encrypt the entire payload for now.

Reply:
I see your point here but I feel like search for a username may be inadvertently looking for a solution to a problem that is not common among users as well as add confusion (and potential information leakage) to how we store variable values.

Reply:
Actually the ability to search for a secret based on certain properties were requested by multiple users.

Reply:
I think that for this kind of search we will have to use some third-party infa for searching in json. We can issue a research for this later

Reply:
Postgres can do that out of the box. We might need to change our column from text to JSON in order for that to work. But that would probably provide the best performance and capabilities.

Reply:
I think that search of a username will be looking for a solution problem